crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1300,7 +1303,15 @@ impl AuthorityPerEpochStore {
                 Some(CmpOrdering::Greater) => false,
                 None => false,
             };
-            checkpoint_service.notify_checkpoint(self, index, roots, final_checkpoint)?;
+            let checkpoint = PendingCheckpoint {
+                roots,
+                details: PendingCheckpointInfo {
+                    timestamp_ms: committed_dag.leader.metadata.created_at,
```

Comment:
What kind of property does this give us regarding the timestamp?
For example, is there a guarantee of non-decreasing?

Reply:
Yes, see comments on the `CheckpointSummary::timestamp_ms` field

Reply:
For my curiosity, how dose Narwhal achieve that? i.e. why wouldn't a byzantine validator be able to create a bad timestamp?

Reply:
Do you think that we should also have some logic here to assert these assumptions, even if they are upheld by narwhal? Eg assert that time is greater than the previous checkpoint and it is not too far in the future from our current clock?

Reply:
I haven't seen the timestamp code, but shouldn't the check be greater than the median (f+1th)? Otherwise dishonest parties would force the time forward

Reply:
@LefKok there is different protection against forcing time forward - primary does not vote if header time is greater then local time, it either discards header(if difference is too large) or sleeps until local time is equal to header time